Source/Core/DiscIO/TGCBlob.cpp

```diff
@@ -0,0 +1,91 @@
+// Copyright 2016 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <memory>
+#include <string>
+
+#include "Common/CommonFuncs.h"
+#include "Common/FileUtil.h"
+#include "DiscIO/TGCBlob.h"
+
+namespace DiscIO
+{
+bool IsTGCBlob(const std::string& path)
+{
+  File::IOFile file(path, "rb");
+
+  TGCHeader header;
+  return (file.ReadArray(&header, 1) && header.magic == Common::swap32(0xAE0F38A2));
+}
+
+TGCFileReader::TGCFileReader(File::IOFile&& file) : m_file(std::move(file))
+{
+  m_file.ReadArray(&m_header, 1);
+  u32 header_size = Common::swap32(m_header.header_size);
+  m_size = m_file.GetSize() - header_size;
+  m_file_offset = Common::swap32(m_header.unknown_important_2) -
+                  Common::swap32(m_header.unknown_important_1) + header_size;
+}
+
+std::unique_ptr<TGCFileReader> TGCFileReader::Create(const std::string& path)
+{
+  File::IOFile file(path, "rb");
+  if (file)
+    return std::unique_ptr<TGCFileReader>(new TGCFileReader(std::move(file)));
+
+  return nullptr;
+}
+
+bool TGCFileReader::Read(u64 offset, u64 nbytes, u8* out_ptr)
+{
+  bool success = true;
+
+  // Offsets before m_file_offset are read as usual
+  if (offset < m_file_offset)
+    success &= InternalRead(offset, std::min(nbytes, m_file_offset - offset), out_ptr);
+
+  // Offsets after m_file_offset are read as if they are (offset - m_file_offset)
+  if (offset + nbytes > m_file_offset)
+    success &= InternalRead(std::max<u64>(offset, m_file_offset) - m_file_offset,
+                            std::min(nbytes, offset + nbytes - m_file_offset), out_ptr);
+
+  return success;
+}
+
+bool TGCFileReader::InternalRead(u64 offset, u64 nbytes, u8* out_ptr)
+{
+  u32 header_size = Common::swap32(m_header.header_size);
+  if (m_file.Seek(offset + header_size, SEEK_SET) && m_file.ReadBytes(out_ptr, nbytes))
+  {
+    Replace32(offset, nbytes, out_ptr, 0x420, SubtractBE32(m_header.dol_offset, header_size));
+    Replace32(offset, nbytes, out_ptr, 0x424, SubtractBE32(m_header.fst_offset, header_size));
+    return true;
+  }
+  else
+  {
+    m_file.Clear();
+    return false;
+  }
+}
+
+u32 TGCFileReader::SubtractBE32(u32 minuend_be, u32 subtrahend_le)
+{
+  return Common::swap32(Common::swap32(minuend_be) - subtrahend_le);
+}
+
+void TGCFileReader::Replace32(u64 offset, u64 nbytes, u8* out_ptr, u32 replace_offset,
+                              u32 replace_value)
+{
+  for (u32 i = 0; i < sizeof(u32); ++i)
+    Replace8(offset, nbytes, out_ptr, replace_offset + i, reinterpret_cast<u8*>(&replace_value)[i]);
+}
+
+void TGCFileReader::Replace8(u64 offset, u64 nbytes, u8* out_ptr, u32 replace_offset,
+                             u8 replace_value)
+{
+  if (offset <= replace_offset && offset + nbytes > replace_offset)
+    out_ptr[replace_offset - offset] = replace_value;
+}
+
+}  // namespace
```

Comment:
For consistency with the other files, maybe there should be a new line at the end of this file?